### PR TITLE
Make pseudo_items less pseudo

### LINKED ISSUE #50289: Electronic Gun Safe cannot be hacked
### Describe the bug

When trying to hack into an electronic gun safe using either an active Fingerhack CBM or an electrohack, receive the message "You don't have a hacking tool with enough charges!"

### Steps To Reproduce

1.  Load save [Pickett.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/6896806/Pickett.zip)
2. Attempt to hack nearby safe

### Expected behavior

Able to use either the CBM or the electrohack to attempt hacking into the safe

### Screenshots

_No response_

### Versions and configuration

- OS: Windows
    - OS Version: 10.0 2009
- Game Version: 2b9d8f5 [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Aftershock [aftershock],
    Magiclysm [magiclysm],
    Blaze Industries [blazeindustries],
    Disable NPC Needs [no_npc_food],
    No Rail Stations [No_Rail_Stations],
    Bionic Professions [package_bionic_professions],
    No Fungal Monsters [No_Fungi]
]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by GoLoT**
#### Summary
None

#### Purpose of change
Fixes #50289

#50326 brought up an issue with the way pseudo_items interact with the inventory system. For tools, pseudo items don't really exist and are created on demand when needed for crafting or specific actions. This makes it hard to work around limitations in some functions that need the items to be valid for an extended period of time (long term activities) or use pointers to those items after they go out of scope.

#### Describe the solution
Refactored the pseudo_item system to create a cache of pseudo items that can be retrieved when needed. It will only be invalidated when a CBM creates or removes a pseudo_item.. This makes pseudo_items behave like regular items where needed (quick-use menu, crafting menu, etc).

Instances of pseudo_items are created and kept inside the `bionic` object that spawns them. For weapons the item will be persisted to the savegame allowing weapons to persist status between save/load cycles. This should allow installed weapons to be modded and with some small changes in code even replaced completely. 

Added the possibility to append the USES_BIONIC_POWER flag to the fake items spawned and modified the finger lighter CBM to use this new feature. It now creates a regular electric lighter item that behaves the same way regular lighters do but instead uses bionic power. This also works for other tools (electrohack), avoiding the need to create new items just to have that extra flag added.

Split fake_item into 3 different fields:
* `passive_pseudo_items` a list of tools that will be available once the CBM is installed.
* `toggled_pseudo_items ` for a list of tools available while the CBM is powered.
* `fake_weapon` the weapon used for gun and weapon CBMs.

Unhardcoded bionic incompatibilities between bio_ears and bio_earplugs and added a new json field `auto_deactivates` that takes a list of CBM ids that will be deactivated automatically when the CBM is activated.

Updated weapon and tool CBMs to use the new fields.

#### Describe alternatives you've considered
Sleeping.

#### Testing
Tested the main tools from CBMs: Electrohack, lighter, integrated toolset, extended toolset.
Also tested weapon CBMs from vanilla and aftershock.

Integrated toolset in crafting menu:
image
image
When out of power:
image

Quick use menu with lighter CBM and enough power:
image
Without power:
image

Fingerhack on an electronic gunsafe:
image

#### Additional context
Follow up PRs/ideas/suggestions:
* Turn integrated toolset into multiple different CBMs with a small set of tools each. Or make it an upgradable CBM with multiple steps, each adding more complex tools to the set.
* Add a new kind of CBM that can hold specific items up to a volume and persists that item. As a first idea: a generic handgun CBM that you can attach a small handgun to and use it like other weapon CBMs. Or a CBM that can have a blade attached with an upgrade path for blades.

### CODE REVIEW

**Comment by GoLoT** [comment]
I've done some basic testing and it's working well so far. I think it's ready for review.

Any feedback and help testing is welcome.

**Comment by Hirmuolio** [comment]
Wouldn't it be simpler to modify the pseudo_items only when installing/uninstalling/activating/deactivating bionics? Fake items are not removed or added ever in other situations.

That would also allow storing long lived data in those items. For example the ammo loaded in the bionic shotgun could just be normal ammo in the fake item.

**Comment by GoLoT** [comment]
> Wouldn't it be simpler to modify the pseudo_items only when installing/uninstalling/activating/deactivating bionics? Fake items are not removed or added ever in other situations.
> 
> That would also allow storing long lived data in those items. For example the ammo loaded in the bionic shotgun could just be normal ammo in the fake item.

I did it that way at first but removing items at the start of the vector invalidated iterators and pointers to items anyway so recreating the vector ended up being cleaner. I also though about keeping all the items in a permanent array and making a vector of pointers to those items for the enabled/available ones, which is more or less your second suggestion. This would have to be a linked list or something so pointers stay valid after removal of CBMs.

The problem with weapon/gun CBMs is that they are assigned (by copy) to the `Character::weapon` field so there is more work involved in keeping state for fake guns. I'll see what can be done.

I'll make some quick changes and see how it works. If it doesn't we can always revert.

Also the item_location code might need updating to add a "from_pseudo_items" class.

File: src/character.h [hunk 1]
```diff
@@ -1234,7 +1234,8 @@ class Character : public Creature, public visitable
         /** Handles bionic activation effects of the entered bionic, returns if anything activated */
         bool activate_bionic( int b, bool eff_only = false, bool *close_bionics_ui = nullptr );
         std::vector<bionic_id> get_bionics() const;
-        std::vector<item> get_pseudo_items() const;
+        std::vector<item> &get_pseudo_items() const;
```
Comment (anothersimulacrum) [inline src/character.h:1237]: I don't think you should be able to freely modify the cached items.
```suggestion
        const std::vector<item> &get_pseudo_items() const;
```

Comment (GoLoT) [reply src/character.h]: Changed it to return a `std::vector<const item*>`, this way we can keep permanent instances of the items and also allowed to add persistence of CBM items (serializing weapons into the savefile).

File: src/visitable.cpp [hunk 2]
```diff
@@ -918,9 +918,8 @@ int Character::amount_of( const itype_id &what, bool pseudo, int limit,
                           const std::function<bool( const item & )> &filter ) const
 {
     if( pseudo ) {
-        for( const auto &bio : *this->my_bionics ) {
-            const bionic_data &bid = bio.info();
-            if( bid.fake_item == what && ( !bid.activated || bio.powered ) ) {
+        for( const auto &pseudos : get_pseudo_items() ) {
```
Comment (anothersimulacrum) [inline src/visitable.cpp:921]:
```suggestion
        for( const item *pseudos : get_pseudo_items() ) {
```

File: src/visitable.cpp [hunk 3]
```diff
@@ -443,8 +443,8 @@ const
         }
     }
 
-    for( const auto &e : get_pseudo_items() ) {
-        if( visit_internal( func, &e ) == VisitResponse::ABORT ) {
+    for( const auto *e : get_pseudo_items() ) {
```
Comment (anothersimulacrum) [inline src/visitable.cpp:446]:
```suggestion
    for( const item *e : get_pseudo_items() ) {
```

File: src/item_action.cpp [hunk 4]
```diff
@@ -247,11 +247,12 @@ void game::item_action_menu()
...
     const action_map &item_actions = gen.get_item_action_map();
 
-    std::vector<item> &pseudo_items = get_player_character().get_pseudo_items();
+    std::vector<const item *> pseudo_items = get_player_character().get_pseudo_items();
     std::vector<item *> pseudos;
     pseudos.reserve( pseudo_items.size() );
-    for( item &pseudo : pseudo_items ) {
-        pseudos.push_back( &pseudo );
+    // Ugly const_cast because the menu needs non-const pointers
+    for( auto &pseudo : pseudo_items ) {
```
Comment (anothersimulacrum) [inline src/item_action.cpp:254]:
```suggestion
    for( const item *pseudo : pseudo_items ) {
```

**Comment by GoLoT** [comment]
Refactored the whole thing again and now each CBM keeps instances of each of the pseudo items it spawns. The list can be retrieved as a `std::vector<const item*>` returning only the pseudo_items for powered/passive CBMs.

Split fake_item into 3 different fields. There is now a `fake_weapon` field and two arrays: `passive_pseudo_items` and `toggled_pseudo_items`.  Passive pseudo items will be available always. Toggled pseudo items will only be available if the CBM is powered.

I've done some testing and nothing broke so far. Saving and loading weapon data seems to be working so far.

**Pull request merged**